pkg/buses/reconciler.go

```diff
@@ -530,36 +538,15 @@ func (r *Reconciler) syncSubscription(namespace string, name string) error {
 	return nil
 }
 
-func (r *Reconciler) createOrUpdateBus(bus *channelsv1alpha1.Bus) error {
-	if r.bus.GetObjectKind().GroupVersionKind().Kind != bus.Kind ||
-		bus.Namespace != r.bus.GetObjectMeta().GetNamespace() ||
-		bus.Name != r.bus.GetObjectMeta().GetName() {
-		// this is not our bus
-		return nil
-	}
-
-	previousBus := r.bus
-	r.bus = bus.DeepCopy()
-	if !equality.Semantic.DeepEqual(r.bus.GetSpec(), previousBus.GetSpec()) {
-		err := r.handler.onBus(r.bus, r)
-		if err != nil {
-			return err
-		}
-	}
-
-	return nil
-}
-
-func (r *Reconciler) createOrUpdateClusterBus(clusterBus *channelsv1alpha1.ClusterBus) error {
-	if r.bus.GetObjectKind().GroupVersionKind().Kind != clusterBus.Kind ||
-		clusterBus.Name != r.bus.GetObjectMeta().GetName() {
-		// this is not our clusterbus
+func (r *Reconciler) createOrUpdateBus(bus channelsv1alpha1.GenericBus) error {
+	if r.ref != NewBusReference(bus) {
+		// not the bus for this reconciler
 		return nil
 	}
 
-	previousBus := r.bus
-	r.bus = clusterBus.DeepCopy()
-	if !equality.Semantic.DeepEqual(r.bus.GetSpec(), previousBus.GetSpec()) {
+	// stash the new bus on the reconciler while retaining the old bus
+	bus, r.bus = r.bus, bus
```

Comment:
Why is this stash needed? How does it work when there are multiple worker threads?

Reply:
The workqueue guarantees that the same resource will never be emitted concurrently, so that should handle any race concerns.

We capture the bus resource for a few reasons:
- in order to resolve parameters for Channels and Subscriptions
- to know if a Channel is backed by the particular bus instance
- to record Events against

Reply:
The same resource will not be emitted concurrently, but IIUC there is still the case of `threadiness` different resources being reconciled simultaneously.

1. resource A enters worker thread A
2. resource B enters worker thread B
3. resource A and B race to set `r.bus`
4. What is the value of `r.bus`?

Is this scenario possible?

Serving and the Feed controller have `context.Context` objects plumbed through the entire Reconcile loop now, and it's reasonable to expect this will be a general pattern in Knative controllers. We may be able to eliminate the race (if there is one) by stashing the bus resource in a `context.Context`. That also makes the single-call scope of the bus stash clear. I'm intentionally ignoring any opinions on the design of `context.Context` for now. :grin:

Reply:
A reconciler is for a single bus resource. Resource A and B will never race to set `r.bus` because only A or B can be for the the bus that is being reconciled (otherwise there's a major bug in workqueue). When `r.bus` is updated, the kind, apiVersion, name and namespace of the previous and future value are identical. We set the value to pickup spec mutations.

#422 adds all provisioned Channels and/or Subscriptions (as needed) to the workqueue when the Bus params change. This will cause them to be reprocessed.

Reply:
@scothis and I talked about this offline. Restating the above a bit: Reconciler will have all Bus update events in its workqueue, but _discards update events for all Buses except one_. That happens at https://github.com/knative/eventing/pull/420/files#diff-d2cf608fb7c41df730d64ac95f391e7aR542.

The Reconciler's workqueue _has multiple object types in it_. This is why the Reconciler needs `threadiness` even though it only ever reconciles a single Bus. This is not conventional. I don't know enough about the implementation to say whether the convention should apply, but I'd to see a comment in the code that acknowledges and briefly explains this break from convention.

So tl;dr my understanding is that this isn't a race condition even though it looks like one. I'd like to see a comment in the code acknowledging that as well.